cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -163,56 +177,269 @@ func (r *Reconciler) spawnOrDeleteMountpointPodIfNeeded(
+	authSource := r.getAuthSource(pv)
+	fsGroup := r.getFSGroup(workloadPod)
+
+	fieldFilters := client.MatchingFields{
+		crdv1.FieldNodeName:             workloadPod.Spec.NodeName,
+		crdv1.FieldPersistentVolumeName: pv.Name,
+		crdv1.FieldVolumeID:             pv.Spec.CSI.VolumeHandle,
+		crdv1.FieldMountOptions:         strings.Join(pv.Spec.MountOptions, ","),
+		crdv1.FieldWorkloadFSGroup:      fsGroup,
+		crdv1.FieldAuthenticationSource: authSource,
+	}
+
+	if authSource == "pod" {
+		fieldFilters[crdv1.FieldWorkloadNamespace] = workloadPod.Namespace
+		fieldFilters[crdv1.FieldWorkloadServiceAccountName] = getServiceAccountName(workloadPod)
+		fieldFilters[crdv1.FieldWorkloadServiceAccountIAMRoleARN] = roleArn
+	}
+
+	return fieldFilters
+}
+
+func (r *Reconciler) getAuthSource(pv *corev1.PersistentVolume) string {
+	volumeAttributes := mppod.ExtractVolumeAttributes(pv)
+	authSource := volumeAttributes[volumecontext.AuthenticationSource]
+	if authSource == "" {
+		return "driver"
+	}
+	return authSource
+}
+
+func (r *Reconciler) getFSGroup(workloadPod *corev1.Pod) string {
+	if workloadPod.Spec.SecurityContext.FSGroup != nil {
+		return strconv.FormatInt(*workloadPod.Spec.SecurityContext.FSGroup, 10)
+	}
+	return ""
+}
+
+func (r *Reconciler) getExistingS3PodAttachments(ctx context.Context, fieldFilters client.MatchingFields) (*crdv1.MountpointS3PodAttachmentList, error) {
+	s3paList := &crdv1.MountpointS3PodAttachmentList{}
+	if err := r.List(ctx, s3paList, fieldFilters); err != nil {
+		return nil, err
+	}
+
+	if len(s3paList.Items) > 1 {
+		return nil, fmt.Errorf("found %d MountpointS3PodAttachments instead of 1", len(s3paList.Items))
+	}
+
+	return s3paList, nil
+}
+
+func (r *Reconciler) handleInactivePod(ctx context.Context, workloadPod *corev1.Pod, s3paList *crdv1.MountpointS3PodAttachmentList, workloadUID string, log logr.Logger) (bool, error) {
+	if len(s3paList.Items) != 1 {
+		log.Info("Workload pod is not active. Did not find any MountpointS3PodAttachments.")
+		return false, nil
+	}
+
+	return r.removeWorkloadFromS3PodAttachment(ctx, &s3paList.Items[0], workloadUID, log)
+}
+
+func (r *Reconciler) handleExistingS3PodAttachment(ctx context.Context, s3paList *crdv1.MountpointS3PodAttachmentList, workloadUID string, fieldFilters client.MatchingFields, log logr.Logger) (bool, error) {
+	s3pa := &s3paList.Items[0]
+
+	if r.s3paExpectations.IsPending(fieldFilters) {
+		log.Info("MountpointS3PodAttachment creation is pending, removing from pending")
+		r.s3paExpectations.Clear(fieldFilters)
+	}
+
+	if s3paContainsWorkload(s3pa, workloadUID) {
+		log.Info("MountpointS3PodAttachment already has this workload UID")
+		return false, nil
+	}
+
+	return r.addWorkloadToS3PodAttachment(ctx, s3pa, workloadUID, log)
+}
+
+func (r *Reconciler) addWorkloadToS3PodAttachment(ctx context.Context, s3pa *crdv1.MountpointS3PodAttachment, workloadUID string, log logr.Logger) (bool, error) {
```

Comment:
Would be nice to have some sanity checks in this function (e.g., checking if `s3pa`s and `workloadPod`'s node, pvName etc. fields match)